flash: enable support for devices without erase

Laczen opened the pull request:
Not all devices that are used with the flash api need to be erased before being written.

~~The PR makes the erase function optional and allows drivers to specify that they have an erase (by selecting `FLASH_HAS_ERASE`), this property is reflected in a Kconfig option `FLASH_ERASE` that can also be selected by subsystems that require an erase.~~

~~The PR adds flags to the `flash_parameters` that is initialized to `FLASH_FLAG_ERASE_OPTIONAL` for devices where the erase is optional.~~

~~The PR provides a new api function `flash_prepare` that is always available and falls back to erase on devices that do not  the flag `FLASH_FLAG_ERASE_OPTIONAL` set. For devices that do have the flag set to `FLASH_FLAG_ERASE_OPTIONAL` the function just returns 0.~~

~~The `flash_prepare` routine can be used as a replacement for `flash_erase` by a subsystem that does not require the area to be erased.~~

~~The subsystem that only needs `flash_prepare` can use the flag to determine that the call to `flash_prepare` can be skipped, but it is left up to the subsystem to use this functionality.~~

~~Devices that do not have an erase function can provide a efficient method to set the erase area to the erase value or they can return `-ENOTSUP`.~~

Reworked the PR to minimize effect on zephyr tree:

The PR adds a Kconfig option that allows drivers to specify that they can write without erase (by selecting `FLASH_HAS_DIRECT_WRITE`). Selecting this Kconfig adds a bool `direct_write` to `flash_parameters`.

Subsystems can find out the value of `direct_write` by using the routine `flash_has_direct_write()`, and adapt their behavior according to this flag.

The system-wide flag `CONFIG_FLASH_HAS_DIRECT_WRITE` is available for subsystems.

The PR redefines `flash_erase()` as placing the flash in a predetermined state (the erase-value). The implementation of `api->erase` is left to the driver but implementing it as a function that returns `-ENOTSUP` could lead to a non-functional subsystem using the flash. It is adviced to implement it (possibly as a set to erased-value) to provide full functionality of all subsystems. As most subsystems will require an erase functionality (either as a real erase or as a means to delete data) the implementation provided by the driver can be optimized to make full use of the driver capabilities.

The PR does not try to avoid the syscall made by `flash_has_direct_write()` on systems that only have devices with direct
write capabilities.

The PR does not require any changes to present subsystems, but it is adviced for subsystems to see if they can take advantage of the direct-write capability.

Remark: The same methodology could be used to reduce the handling of the flash_parameter `erase-value` that is 0xFF on most systems.

The PR adds a configuration option to the flash simulator to simulate direct-write devices.

The PR shows how subsystems (i.e. flash_disk) can take advantage of the direct-write property.

The PR is an alternative to #71271.

Laczen committed:
flash: add support for devices without erase

Not all devices that are used with the flash api need to be erased
before being written.

The PR adds a Kconfig option that allows drivers to specify that they
can write without erase (by selecting `FLASH_HAS_DIRECT_WRITE`).
Selecting this Kconfig adds a bool `direct_write` to `flash_parameters`.

Subsystems can find out the value of `direct_write` by using the
routine `flash_has_direct_write()`, and adapt their behavior according
to this flag.

The system-wide flag `CONFIG_FLASH_HAS_DIRECT_WRITE` is available for
subsystems.

The PR redefines `flash_erase()` as placing the flash in a predetermined
state (the erase-value). The implementation of `api->erase` is left to
the driver but implementing it as a function that returns `-ENOTSUP`
could lead to a non-functional subsystem using the flash. It is adviced
to implement it (possibly as a set to erased-value) to provide full
functionality of all subsystems.

The PR does not try to avoid the syscall made by
`flash_has_direct_write()` on systems that only have devices with direct
write capabilities.

The PR does not require any changes to present subsystems, but it is
adviced for subsystems to see if they can take advantage of the direct-
write capability.

Remark: The same methodology could be used to reduce the handling of the
flash_parameter `erase-value` that is 0xFF on most systems.

The PR is an alternative to #71271.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen commented:
@carlescufi, @henrikbrixandersen, @bjarki-trackunit, @jfischer-no, @nordicjm, FYI.

Laczen commented:
@RichardSWheatley, @aaronyegx, FYI.

carlescufi reviewed (commented):
@Laczen please include examples on how higher-level subsystems would use `flash_prepare` and the new `flags` to behave optimally. @de-nordic has provided that in https://github.com/zephyrproject-rtos/zephyr/pull/71271

de-nordic commented:
This is not alternative to https://github.com/zephyrproject-rtos/zephyr/pull/71271, it barely touches the problem.
The main goal of the original was to allow subsystems to identify capability of devices (need to erase) and work with them (erase if required by design, do something else if not), and show how it will be done. This implementation is completely superficial and that is why is fails in its basics.

Main issue, **and the one that describes the lack of understanding of the original  approach in #71271 and the problem it solves**, is with the design and implementation of the `flash_prepare` that  returns:
 - `0` when erase not called because `CONFIG_FLASH_ERASE=n`
 - `0` when `CONFIG_FLASH_ERASE=y` and `((flags & FLASH_FLAG_ERASE_OPTIONAL) == FLASH_FLAG_ERASE_OPTIONAL)`
 - `0` when `api->erease` returns `0`.
 
**This makes the function useless/pointless in common code working with different devices, because once you call it you do not even know whether the device you currently work with has been "prepared" for anything.**
You get 0 (success) and do not know what it means without checking the flag.

Every sane subsystem will end up just checking the flag and calling the `flash_erase` or doing non-erase preparations by hand, depending on what subsystem deems as preparation in its case. This is the point where this PR slowly turns to #71271.

I have answered request for similar function here: https://github.com/zephyrproject-rtos/zephyr/pull/71271#issuecomment-2102789040, a few comments before you have blocked my PR.

The main, overall goal, that follows the #71271, is to remove erase callback where not needed, not make it optional where not needed.

Erase would be optional if MRAM/FRAM/RRAM drivers could do `api->erase = NULL`; this PR makes erase implementation mandatory (when `CONFIG_FLASH_ERASE=y`) or not-available  (when `CONFIG_FLASH_ERASE=n`) for all devices.

Laczen commented:
> @Laczen please include examples on how higher-level subsystems would use `flash_prepare` and the new `flags` to behave optimally. @de-nordic has provided that in #71271

@carlescufi there is no need to use the flags to behave optimally, because:
1. The saved cpu cycles would be very little,
2. This requires extra code to retrieve the flag,
3. It requires extra code to handle the flag,
4. It will only appear when there is a combination of flash with erase and without erase requirements, and even then only when subsystem will be running on the flash without erase requirements.

As 4 will only appear in very specific conditions while 2 and 3 would appear for all solutions the "optimization" would not be a valid optimisation.

de-nordic commented:
> 
> The function `flash_prepare()` is short for `flash_prepare_for_write()`, in all the above returns of 0 it means that the flash area can be written.

There is no need for a function to prepare for write, we need a function to substitute where erase is not available and has been used due to subsystem design or as it was available way to remove data.
Semantics of preparing something for write depends on subsystem, devices used by that subsystem will never know or understand what subsystem tries to do, they will understand what they are supposed to do according to their design and interface they provide.

> 
> When the function returns 0 flash can be written, no need to check any flag.

No need to call it, go for flag and erase.

>
> No they will just use flash when the function returned that it can be written (`flash_prepare()` returned 0).

There is no reason to call the function, the entire premise of its existence is pointless.

>
> The function `flash_prepare()` is indeed very similar to `flash_erase_when_needed()`.

So again  [#71271 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/71271#issuecomment-2102789040)
Another thing is that how to prepare device for write depends on subsystem that uses the device not on device, the device only makes it easier or harder.
For example writing 32 bytes of data requires different preparation when you can just write or need to erase first, this preparation is not limited to device itself but also extends to memory allocation and solving problems like what to do with the rest of data that will be erased when only 32 bytes are needed. 

> 
> There are certain rules we have to follow for the return values of api functions (https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/api/design_guidelines.rst?plain=1#L106C1-L122C57).
>
> Optional API functions that are not implemented should return `-ENOSUP`. This could be done at the `flash.h` level in `flash_erase()`, 

Yeah, the https://github.com/zephyrproject-rtos/zephyr/pull/71271 does that for a long time now.

> but a more efficient solution is to provide a driver api->erase that just returns `-ENOTSUP`. A solution in `flash.h` would require a check for NULL that is not needed when all drivers do provide a api->erase.

This is how null check looks on arm
![image](https://github.com/zephyrproject-rtos/zephyr/assets/56024351/0d380aa0-045e-4be5-a6fd-20f14820ff9b)
this is how `flash_erase` calls look in the same binary:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/56024351/c2478ae6-095e-4201-b0cd-0a4afaa6872c)

**On the other side each driver call that only returns -ENOTSUP will consist of prolog, r0 assignment and apilog ~7 ops.**

In both cases checking for `-ENOTSUP` is pointless, on return, because it means that call has been done for no reason.
This is not *NIX where compiled version of software has to account for various hardware configurations - hardware configuration is constant for a system, devices have something or they do not, and that is known at compile time.

de-nordic commented:
> 
> @carlescufi there is no need to use the flags to behave optimally, because:
> 
>     1. The saved cpu cycles would be very little,

I have adapted subsystems to work most efficiently with what device offers.
Erasing simulated 4k page may be ~4k CPU cycles for nothing in case when not needed, this also involves internal bus, power, etc. And not everyone is happy with wasted resources.

> 
>     2. This requires extra code to retrieve the flag,

The code is there, mostly, because subsystems have to align to device properties. And checking the flag only happens on devices that are able to use different logic for erase requiring and non-erase devices.
Subsystems designed around erase have to substitute for erase unconditionally, unless redesigned.

> 
>     3. It requires extra code to handle the flag,

The flag does not matter, what matters is how the system addresses it. There is often completely different approach needed when the flag exists and when it does not. Checking a flag is dime-a-dozen when the supporting code takes most of flash/cpu.

> 
>     4. It will only appear when there is a combination of flash with erase and without erase requirements, and even then only when subsystem will be running on the flash without erase requirements.

Code that calls the function still does not know whether it works with device requiring erase or not, and has to check the flag to figure out how to write, what to write and how manage data on device.
The function call is pointless.

> 
> 
> As 4 will only appear in very specific conditions while 2 and 3 would appear for all solutions the "optimization" would not be a valid optimisation.
> 
> Real optimization for subsystems would require build time knowledge on the flash memory type that is user. Build time knowledge would avoid any runtime calls to retrieve the memory type/properties and still allow selecting the correct action to be performed. This information could e.g. be retrieved from the dts file.

So why do you propose broken design in the first place?
By your own words you know that this is bad idea or bad design, yet you are proposing it as alternative.

You have no clear or consistent approach to the problem.

And once more: DTS properties should not lead to lasagna-code anti-pattern, which will happen if parameters for subsystems end up in device definitions. I have already explained that https://github.com/zephyrproject-rtos/zephyr/files/14795769/QA.no-erase-flash.pdf , 9 and 10

Laczen commented:
> > The function `flash_prepare()` is short for `flash_prepare_for_write()`, in all the above returns of 0 it means that the flash area can be written.
> 
> There is no need for a function to prepare for write, we need a function to substitute where erase is not available and has been used due to subsystem design or as it was available way to remove data. Semantics of preparing something for write depends on subsystem, devices used by that subsystem will never know or understand what subsystem tries to do, they will understand what they are supposed to do according to their design and interface they provide.
> 
> > When the function returns 0 flash can be written, no need to check any flag.
> 
> No need to call it, go for flag and erase.
> 
> > No they will just use flash when the function returned that it can be written (`flash_prepare()` returned 0).
> 
> There is no reason to call the function, the entire premise of its existence is pointless.
> 
> > The function `flash_prepare()` is indeed very similar to `flash_erase_when_needed()`.
> 
> So again [#71271 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/71271#issuecomment-2102789040) Another thing is that how to prepare device for write depends on subsystem that uses the device not on device, the device only makes it easier or harder. For example writing 32 bytes of data requires different preparation when you can just write or need to erase first, this preparation is not limited to device itself but also extends to memory allocation and solving problems like what to do with the rest of data that will be erased when only 32 bytes are needed.
> 

Clearly from the comment the API users are not interested in what is going on behind the scene, the user would like a new area for writing. A good API is simple to understand and hides any implementation details from the user. The user is not interested in finding out what API routine it has to call or skip, the user just wants a simple routine that does what is needed.

It is clear from the documentation what limitations apply to the call, it is not intended as a preparation for whatever size. It will be allowed to call the preparation with a limited size on some devices, but not on most.

> > There are certain rules we have to follow for the return values of api functions (https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/api/design_guidelines.rst?plain=1#L106C1-L122C57).
> > Optional API functions that are not implemented should return `-ENOSUP`. This could be done at the `flash.h` level in `flash_erase()`,
> 
> Yeah, the #71271 does that for a long time now.
>
Yeah, and it will fail with a KERNEL_OOPS.
 
> > but a more efficient solution is to provide a driver api->erase that just returns `-ENOTSUP`. A solution in `flash.h` would require a check for NULL that is not needed when all drivers do provide a api->erase.
> 
> This is how null check looks on arm ![image](https://private-user-images.githubusercontent.com/56024351/332378903-0d380aa0-045e-4be5-a6fd-20f14820ff9b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyOTQ4NjAsIm5iZiI6MTcxNjI5NDU2MCwicGF0aCI6Ii81NjAyNDM1MS8zMzIzNzg5MDMtMGQzODBhYTAtMDQ1ZS00YmU1LWE2ZmQtMjBmMTQ4MjBmZjliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIxVDEyMjkyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmZWRkNjM1MGQ1MWRmMTZkOWQ5OTVkOTNmMWM1ZTEwYjZmODVlNThmNGRlOGNjNjUzOWI2ODBkMjc5NzhmMjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.brrVKQOJXXDFhJ21veKtw6ztyO4xqu7QIWQT_jCuEMo) this is how `flash_erase` calls look in the same binary: ![image](https://private-user-images.githubusercontent.com/56024351/332378146-c2478ae6-095e-4201-b0cd-0a4afaa6872c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyOTQ4NjAsIm5iZiI6MTcxNjI5NDU2MCwicGF0aCI6Ii81NjAyNDM1MS8zMzIzNzgxNDYtYzI0NzhhZTYtMDk1ZS00MjAxLWIwY2QtMGE0YWZhYTY4NzJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIxVDEyMjkyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNmZTUwMmVmZDFmZTIxOTJmMTMyYTYzZTQ0N2E2MjNjZTNlYTI4ODgwODJjZjQ2MTEyNDdmOTRhNDY2YjUwZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lYZh_4Em-SgT9CMCmJZRPOURnxvoab14j8bK0imy3pU)
> 
> **On the other side each driver call that only returns -ENOTSUP will consist of prolog, r0 assignment and apilog ~7 ops.**
> 
> In both cases checking for `-ENOTSUP` is pointless, on return, because it means that call has been done for no reason. This is not *NIX where compiled version of software has to account for various hardware configurations - hardware configuration is constant for a system, devices have something or they do not, and that is known at compile time.

The check for NULL will add code size to **ALL** systems of which the majority (>95%?) will not need it.

Laczen commented:
> > @carlescufi there is no need to use the flags to behave optimally, because:
> > ```
> > 1. The saved cpu cycles would be very little,
> > ```
> 
> I have adapted subsystems to work most efficiently with what device offers. Erasing simulated 4k page may be ~4k CPU cycles for nothing in case when not needed, this also involves internal bus, power, etc. And not everyone is happy with wasted resources.
> 
It seems that you don't understand what extra cpu cycles are involved, I have added some extra documentation to the include file to make it clear(er) what `flash_prepare()` does:
```
 *  @brief  Prepare part or all of a flash memory for writing
 *
 *  The flash memory area is prepared for writing by either calling erase
 *  (on devices that need to be erased prior to writing) or by doing nothing,
 *  returning 0 (on devices that do not need to be erased prior to writing).
```
The cpu cycles that are involved are the switch to the system thread (when userspace is used) just to return 0 based upon the internal flag, there is no erasing being done.
 
> > ```
> > 2. This requires extra code to retrieve the flag,
> > ```
> 
> The code is there, mostly, because subsystems have to align to device properties. And checking the flag only happens on devices that are able to use different logic for erase requiring and non-erase devices. Subsystems designed around erase have to substitute for erase unconditionally, unless redesigned.
>

Subsystems shouldn't be checking flags and adapting there behavior to flags. As shown by `flash_disk` they are not using any device properties.
 
> > ```
> > 3. It requires extra code to handle the flag,
> > ```
> 
> The flag does not matter, what matters is how the system addresses it. There is often completely different approach needed when the flag exists and when it does not. Checking a flag is dime-a-dozen when the supporting code takes most of flash/cpu.
>
Again, there shouldn't be any flag checking and adapting there behavior.
 
> > ```
> > 4. It will only appear when there is a combination of flash with erase and without erase requirements, and even then only when subsystem will be running on the flash without erase requirements.
> > ```
> 
> Code that calls the function still does not know whether it works with device requiring erase or not, and has to check the flag to figure out how to write, what to write and how manage data on device. The function call is pointless.
> 
> > As 4 will only appear in very specific conditions while 2 and 3 would appear for all solutions the "optimization" would not be a valid optimisation.
> > Real optimization for subsystems would require build time knowledge on the flash memory type that is user. Build time knowledge would avoid any runtime calls to retrieve the memory type/properties and still allow selecting the correct action to be performed. This information could e.g. be retrieved from the dts file.
> 
> So why do you propose broken design in the first place? By your own words you know that this is bad idea or bad design, yet you are proposing it as alternative.
> 
> You have no clear or consistent approach to the problem.
>
There is nothing broken or inconsistency, adding extra code (for all systems) to be able to "optimize" in one specific case is not optimal at all. It is far more optimal for zephyr as a whole not to optimize this specific case (especially given its minor effect).
  
> And once more: DTS properties should not lead to lasagna-code anti-pattern, which will happen if parameters for subsystems end up in device definitions. I have already explained that https://github.com/zephyrproject-rtos/zephyr/files/14795769/QA.no-erase-flash.pdf , 9 and 10

As you know I completely disagree on this explication regarding what should and what shouldn't be in the dts. What we add in the dts can be triggered by the need for the build system (e.g. the write-block-size that is used for alignment in `west sign`, or for a different reason, e.g. to allow build time optimizations.

Laczen commented:
> > It seems that you don't understand what extra cpu cycles are involved, I have added some extra documentation to the include file to make it clear(er) what `flash_prepare()` does:
> > ```
> >  *  @brief  Prepare part or all of a flash memory for writing
> >  *
> >  *  The flash memory area is prepared for writing by either calling erase
> >  *  (on devices that need to be erased prior to writing) or by doing nothing,
> >  *  returning 0 (on devices that do not need to be erased prior to writing).
> > ```
> > The cpu cycles that are involved are the switch to the system thread (when userspace is used) just to return 0 based upon the internal flag, there is no erasing being done.
> 
> Omg, all the time I was talking about the switch it takes to make the call: https://docs.zephyrproject.org/latest/samples/userspace/syscall_perf/README.html the flag check, to not erase, **happens on the kernel side of syscall, after jumping through the syscall gates**. This is needless cost if you can check the flag and avoid calling the function in the first case. This is how `flash_erase` looks like, and what is called when `CONFIG_USERSPACE=y` ![image](https://private-user-images.githubusercontent.com/56024351/332461260-8e974fcf-5b9d-457e-939f-b52f2c4994b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzMTMyMjUsIm5iZiI6MTcxNjMxMjkyNSwicGF0aCI6Ii81NjAyNDM1MS8zMzI0NjEyNjAtOGU5NzRmY2YtNWI5ZC00NTdlLTkzOWYtYjUyZjJjNDk5NGIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIxVDE3MzUyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJlOTI3MzBlNjUyMWZkMmI0ZTU0Mzg2OTI2ZGY0NmQ3MjAxZDk1ZWE5MmU5MzA0ZjdlZGY1ZjBmNjQ2NWM5ODUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Q5MX4ZEqqN7H2nDjCOQFlq3qZznEiQfjwt1CJGiOyy4) The syscall takes `z_mrsh_flash_erase` from dispatch table, by id `K_SYSCALL_FLASH_ERASE`, the sequence of call looks like: `flash_erase()->syscall(K_SYSCALL_FLASH_ERASE)->z_mrsh_flash_erase()->z_vrfy_flash_erase()-z_impl_flash_erase()->(api->erase())`
> 
> In your case replace `flash_erase` with `flash_prepare` and you did all the path to return zero **after checking the flag**. In worst case scenario you do the `api-><fun>` call twice per syscall.
>
> > It seems that you don't understand what extra cpu cycles are involved,

I'm sorry, but from your comment that a simulated flash_erase takes 4K cycles, I concluded that you were thinking that a call to api->erase would be made.

On devices that do not require erase but is enabled on a system that has erase (either by Kconfig setting or by driver selection) the extra cpu cycles would be:
`flash_prepare() -> syscall(K_SYSCALL_FLASH_PREPARE) -> z_mrsh_flash_prepare_sc() -> z_vfry_flash_prepare_sc() -> z_impl_flash_prepare(), do flag check and return 0`.
These extra cycles will only appear in very specific conditions. Adding methods and code for all systems to optimize this specific condition is not optimal.
> 
> **You have no idea how system calls work**.

This is a rude comment and not inline with the zephyr code of conduct (https://github.com/zephyrproject-rtos/zephyr/blob/main/CODE_OF_CONDUCT.md). 
> 
> > As you know I completely disagree on this explication regarding what should and what shouldn't be in the dts. What we add in the dts can be triggered by the need for the build system (e.g. the write-block-size that is used for alignment in `west sign`, or for a different reason, e.g. to allow build time optimizations.
> 
> This allows optimizations only to subsystems that have single instance, and could be tuned with Kconfig as well. We have what we have now. Once again, I do not see here the DTS approach here, to show how that could work out.
> 

When devices have the property "erase-not-required" set in the devicetree a subsystem can find out this property at compile time, there is no need for a runtime (system)call. The same "optimization" can be done as in your proposal, but it is done at compile time (and without any runtime call).
 
> I do not see how various subsystems work when `CONFIG_FLASH_ERASE=n` is set.

Subsystems that require erase select `CONFIG_FLASH_ERASE=y`, as they need support for flash erase. One example of such a system is nvs. When this is not selected an error will occur as the `flash_erase()` is not available. When it is selected but nvs would be using a device that has `api->erase() == return -ENOTSUP` the system will fail the first time `flash_erase()` is called, just as it is now.

NVS should evolve into something that does not need `flash_erase()` but use `flash_prepare()` instead, in that case there will be no need to select `CONFIG_FLASH_ERASE=y`.

henrikbrixandersen commented:
> When devices have the property "erase-not-required" set in the devicetree a subsystem can find out this property at compile time, there is no need for a runtime (system)call. The same "optimization" can be done as in your proposal, but it is done at compile time (and without any runtime call).

The subsystem, that needs to inspect this property is the flash driver subsystem, not each and every subsystem wanting to use the flash driver subsystem. The former is proper encapsulation, the latter is - apart from a being a poor design - a maintenance nightmare.

Laczen commented:
> > When devices have the property "erase-not-required" set in the devicetree a subsystem can find out this property at compile time, there is no need for a runtime (system)call. The same "optimization" can be done as in your proposal, but it is done at compile time (and without any runtime call).
> 
> The subsystem, that needs to inspect this property is the flash driver subsystem, not each and every subsystem wanting to use the flash driver subsystem. The former is proper encapsulation, the latter is - apart from a being a poor design - a maintenance nightmare.

To be clear this is not proposed in the present PR, it could be used to create a optimized subsystem that has minimal code size and no unneeded syscalls.

de-nordic commented:
> 
> On devices that do not require erase but is enabled on a system that has erase (either by Kconfig setting or by driver selection) the extra cpu cycles would be: `flash_prepare() -> syscall(K_SYSCALL_FLASH_PREPARE) -> z_mrsh_flash_prepare_sc() -> z_vfry_flash_prepare_sc() -> z_impl_flash_prepare(), do flag check and return 0`. These extra cycles will only appear in very specific conditions. Adding methods and code for all systems to optimize this specific condition is not optimal.

Care to explain the conditions? I guess that is on me, again.
So generally, fellow Zephyr users and vendors with MRAM/FRAM/RRAM devices, if your customer has only your SoC in design then yes, if user calls `flash_prepare` in code, you have:
```
  flash_prepare()>return 0 // or even optimized out
```
similar as the https://github.com/zephyrproject-rtos/zephyr/pull/71271 is able to do, but what if you user adds external flash to design, like mx25, m25, 25q? Well then `CONFIG_FLASH_ERASE=y` turns the above sequence to:
```
flash_prepare() -> syscall(K_SYSCALL_FLASH_PREPARE) -> z_mrsh_flash_prepare_sc() -> z_vfry_flash_prepare_sc() -> z_impl_flash_prepare()->return 0
```

It is a bad design if function that is, by design, suppose to do nothing suddenly spikes in CPU when unrelated subsystem or device is enabled. The spikes will happen each time the function is passed, where it was previously doing nothing.

In short, the condition is: enabling unrelated subsystem using external memory with erase requirements will take great cpu impact on subsystems that may not be even using the memory device, but have been calling the `flash_prepare()`, on some other device, every time the function is passed by code execution.

Above happens when you have `CONFIG_USERSPACE=y`. Details here, but should be obvious by the call path presented above.
https://docs.zephyrproject.org/latest/samples/userspace/syscall_perf/README.html

How is checking of a flag, which takes a up to a few cycles when cached, to choose an optimal execution path worse than going full syscall for the same purpose with needless call to the function each time?

There is also seem to be a confusion what optimal and optimized means.
Optimal path in utilizing a device is where a software recognizes hardware features and uses them to its advantage - this is good design; example of non-optimal solution: enforcing erase on a devices that have no such function instead of adapting software to work with such devices.
Optimization is reducing needless operations. For example caching information queried from a device, instead of doing query each time,  when such information never changes,  or removing code branches that will never be taken with given hardware.

mmahadevan108 commented on drivers/flash/Kconfig.at45:
```diff
@@ -23,7 +24,7 @@ if SPI_FLASH_AT45
 config SPI_FLASH_AT45_USE_READ_MODIFY_WRITE
 	bool "Use Read-Modify-Write command in flash_write()"
-	default y
+	default n
```
Why has this changed?

Laczen replied:
This was changed because it could result in faster wear of these devices. I can change it back and threat these devices as devices without erase when this config is selected.

mmahadevan108 replied:
I think that would make sense to conditionalize the erase based on this config. However this does not seem to be related to the intent behind this PR. So probably make this change a separate commit.

Laczen commented:
> Care to explain the conditions? I guess that is on me, again. So generally, fellow Zephyr users and vendors with MRAM/FRAM/RRAM devices, if your customer has only your SoC in design then yes, if user calls `flash_prepare` in code, you have:
>
This remark is needlessly insulting and incorrect, it was clearly stated from the start in which conditions this would occur. 

> There is also seem to be a confusion what optimal and optimized means. Optimal path in utilizing a device is where a software recognizes hardware features and uses them to its advantage - this is good design; example of non-optimal solution: enforcing erase on a devices that have no such function instead of adapting software to work with such devices. Optimization is reducing needless operations. For example caching information queried from a device, instead of doing query each time, when such information never changes, or removing code branches that will never be taken with given hardware.

An optimal solution for zephyr as a whole will take into consideration required cpu cycles but also code size in all configurations. A solution that caches information by querying info from a device might allow a reduction in needless operation in a specific solution but forces all systems to cache this information even in configurations where this doesn't matter. The caching of information is a needless operation that adds code size to configurations that don't need it.

There are also other ways to achieve an optimized (optimal) solution. The flash users could be "placed in the driver seat" and specify how they would like to use the flash subsystem (e.g. specifying that they will skip erase), a "runtime verification" could be included (e.g. in debug build) to verify that the specified method is valid. This would allow removing any unncessary (syscall) routines from the normal build.

de-nordic commented:
> 
> This remark is needlessly insulting and incorrect, it was clearly stated from the start in which conditions this would occur.

Sorry you feel that way I do not find the remark neither insulting nor incorrect. I actually voice here how painful it is that in this, now one week old review, I am the only one to go deep into details what this PR does provide, analyze its effects on subsystems and get back with reporting them, I guess that is expected from me by the role - finding significant problems with code. Unfortunately I can not find neither in the PR description, nor in any review comments, something explaining conditions and the effects that will, truthfully, state:

_with CONFIG_USERSPACE=y when you enable device with erase, i.e CONFIG_FLASH_ERASE=y, unrelated subsystem that has not been even using the device, but runs on xRAM device, will take spike in CPU usage every time it passes `flash_prepare`, even though the function, by design, is supposed to do nothing on such device._

nevertheless I welcome to be corrected with link to such comment, in which case I stand corrected and I am sorry I missed it. I hope I have not wasted time on the analysis.

> 
> An optimal solution for zephyr as a whole will take into consideration required cpu cycles but also code size in all configurations. A solution that caches information by querying info from a device might allow a reduction in needless operation in a specific solution but forces all systems to cache this information even in configurations where this doesn't matter. The caching of information is a needless operation that adds code size to configurations that don't need it.

It forces nobody: it gives an opportunity. If any subsystem is able to take opportunity in caching flags and **not** doing call, that system is free to do that. That system is also to not do that and waste cycles on needless calls, but I guess 90% of devs will choose to former over the later, because they have to query device for other parameters anyway.

The `flash_prepare` has a side effect that takes significant hit on innocent subsystem, that are needless operations by definition.

> 
> There are also other ways to achieve an optimized (optimal) solution. The flash users could be "placed in the driver seat" and specify how they would like to use the flash subsystem (e.g. specifying that they will skip erase). (...)

Checking a flag before doing needless call is, by definition, being "placed in the driver seat".

Laczen commented:
> > This remark is needlessly insulting and incorrect, it was clearly stated from the start in which conditions this would occur.
> 
> Sorry you feel that way I do not find the remark neither insulting nor incorrect. I actually voice here how painful it is that in this, now one week old review, I am the only one to go deep into details what this PR does provide, analyze its effects on subsystems and get back with reporting them, I guess that is expected from me by the role - finding significant problems with code. Unfortunately I can not find neither in the PR description, nor in any review comments, something explaining conditions and the effects that will, truthfully, state:
> 
> _with CONFIG_USERSPACE=y when you enable device with erase, i.e CONFIG_FLASH_ERASE=y, unrelated subsystem that has not been even using the device, but runs on xRAM device, will take spike in CPU usage every time it passes `flash_prepare`, even though the function, by design, is supposed to do nothing on such device._
> 
> nevertheless I welcome to be corrected with link to such comment, in which case I stand corrected and I am sorry I missed it. I hope I have not wasted time on the analysis.

https://github.com/zephyrproject-rtos/zephyr/pull/72924#issuecomment-2119314047 item 4. is exactly that condition. 
> 
> > An optimal solution for zephyr as a whole will take into consideration required cpu cycles but also code size in all configurations. A solution that caches information by querying info from a device might allow a reduction in needless operation in a specific solution but forces all systems to cache this information even in configurations where this doesn't matter. The caching of information is a needless operation that adds code size to configurations that don't need it.
> 
> It forces nobody: it gives an opportunity. If any subsystem is able to take opportunity in caching flags and **not** doing call, that system is free to do that. That system is also to not do that and waste cycles on needless calls, but I guess 90% of devs will choose to former over the later, because they have to query device for other parameters anyway.
> 
> The `flash_prepare` has a side effect that takes significant hit on innocent subsystem, that are needless operations by definition.
> 
> > There are also other ways to achieve an optimized (optimal) solution. The flash users could be "placed in the driver seat" and specify how they would like to use the flash subsystem (e.g. specifying that they will skip erase). (...)
> 
> Checking a flag before doing needless call is, by definition, being "placed in the driver seat".

No, "placed in the driver seat" means that the subsystem will e.g. state: "flash, I will use you without erase", a run time check can error out if this is not ok.

de-nordic commented:
> 
> [#72924 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/72924#issuecomment-2119314047) item 4. is exactly that condition.

OK, I stand half-corrected, as the fact that from doing nothing in no or few instructions it goes to doing nothing but in hundreds of cycles it not emphasized.
Checking the flag goes from noting in no or few instructions to action or another, then there is the cost of action or none.

Still this will happen to significant percentage of users, in double digits.

> 
> No, "placed in the driver seat" means that the subsystem will e.g. state: "flash, I will use you without erase"

Subystem should blindly do so and check what happens? That is not considered good driving strategy.

>  a run time check can error out if this is not ok.

I do not get that part. So:
  - checking the  flag to avoid path with error, not wasting cycles on syscall  --  is not is not ok and waste of cycles
  - needless use of a function (which could be avoided by checking the flag first), that wastes cycles on doing nothing, and the checking eventual error on out -- is ok and not waste of cycles.

People have literally invented cache in hardware to avoid taking slowest path of fetching data again, this goes literally the opposite way.

Laczen committed:
flash_simulator: add direct write support

Added support for the flash simulator to behave as flash with direct
write (no erase required before write)

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen committed:
flash_disk: support flash with direct write

Modify flash_disk to take advantage of direct write.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen commented:
> > [#72924 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/72924#issuecomment-2119314047) item 4. is exactly that condition.
> 
> OK, I stand half-corrected, as the fact that from doing nothing in no or few instructions it goes to doing nothing but in hundreds of cycles it not emphasized. Checking the flag goes from noting in no or few instructions to action or another, then there is the cost of action or none.
> 
> Still this will happen to significant percentage of users, in double digits.
>

This is no longer relevant for the updated PR.
 
> > No, "placed in the driver seat" means that the subsystem will e.g. state: "flash, I will use you without erase"
> 
> Subystem should blindly do so and check what happens? That is not considered good driving strategy.
> 
> > a run time check can error out if this is not ok.
> 
> I do not get that part. So:
> 
> * checking the  flag to avoid path with error, not wasting cycles on syscall  --  is not is not ok and waste of cycles
> * needless use of a function (which could be avoided by checking the flag first), that wastes cycles on doing nothing, and the checking eventual error on out -- is ok and not waste of cycles.
> 
> People have literally invented cache in hardware to avoid taking slowest path of fetching data again, this goes literally the opposite way.

At the moment we are adding code to "production" releases that are retrieving *constant* values such as write-block-size, erase-block-size, the ability to write without prior erase, ... All these things never change but as part of an init sequence they are retrieved (and many by a syscall). An alternative approach would be to set these values by the subsystem, use a "validation" release that verifies the set values. The "production" release can then be relieved of all retrieving routines.
